packages/storage-plus/src/snapshot/mod.rs

```diff
@@ -0,0 +1,377 @@
+
+        assert_eq!(NEVER.has_changelog(&mut storage, DUMMY_KEY, 3), Ok(false));
+        assert_eq!(EVERY.has_changelog(&mut storage, DUMMY_KEY, 3), Ok(false));
+        assert_eq!(SELECT.has_changelog(&mut storage, DUMMY_KEY, 3), Ok(false));
+
+        // Write a changelog at 2
+        NEVER
+            .write_changelog(&mut storage, DUMMY_KEY, 2, Some(3))
+            .unwrap();
+        EVERY
+            .write_changelog(&mut storage, DUMMY_KEY, 2, Some(3))
+            .unwrap();
+        SELECT
+            .write_changelog(&mut storage, DUMMY_KEY, 2, Some(3))
+            .unwrap();
+
+        assert_eq!(NEVER.has_changelog(&mut storage, DUMMY_KEY, 1), Ok(false));
+        assert_eq!(EVERY.has_changelog(&mut storage, DUMMY_KEY, 1), Ok(false));
+        assert_eq!(SELECT.has_changelog(&mut storage, DUMMY_KEY, 1), Ok(false));
+
+        assert_eq!(NEVER.has_changelog(&mut storage, DUMMY_KEY, 2), Ok(true));
+        assert_eq!(EVERY.has_changelog(&mut storage, DUMMY_KEY, 2), Ok(true));
+        assert_eq!(SELECT.has_changelog(&mut storage, DUMMY_KEY, 2), Ok(true));
+
+        assert_eq!(NEVER.has_changelog(&mut storage, DUMMY_KEY, 3), Ok(false));
+        assert_eq!(EVERY.has_changelog(&mut storage, DUMMY_KEY, 3), Ok(false));
+        assert_eq!(SELECT.has_changelog(&mut storage, DUMMY_KEY, 3), Ok(false));
+    }
+
+    #[test]
+    fn may_load_at_height() {
+        let mut storage = MockStorage::new();
+
+        assert_eq!(
+            NEVER.may_load_at_height(&storage, DUMMY_KEY, 3),
+            Err(StdError::not_found("checkpoint"))
+        );
+        assert_eq!(EVERY.may_load_at_height(&storage, DUMMY_KEY, 3), Ok(None));
+        assert_eq!(
+            SELECT.may_load_at_height(&storage, DUMMY_KEY, 3),
+            Err(StdError::not_found("checkpoint"))
+        );
+
+        // Add a checkpoint at 3
+        NEVER.add_checkpoint(&mut storage, 3).unwrap();
+        EVERY.add_checkpoint(&mut storage, 3).unwrap();
+        SELECT.add_checkpoint(&mut storage, 3).unwrap();
+
+        assert_eq!(
+            NEVER.may_load_at_height(&storage, DUMMY_KEY, 3),
+            Err(StdError::not_found("checkpoint"))
+        );
+        assert_eq!(EVERY.may_load_at_height(&storage, DUMMY_KEY, 3), Ok(None));
+        assert_eq!(SELECT.may_load_at_height(&storage, DUMMY_KEY, 3), Ok(None));
+
+        // Write a changelog at 3
+        NEVER
+            .write_changelog(&mut storage, DUMMY_KEY, 3, Some(100))
+            .unwrap();
+        EVERY
+            .write_changelog(&mut storage, DUMMY_KEY, 3, Some(100))
+            .unwrap();
+        SELECT
+            .write_changelog(&mut storage, DUMMY_KEY, 3, Some(100))
+            .unwrap();
+
+        assert_eq!(
+            NEVER.may_load_at_height(&storage, DUMMY_KEY, 3),
+            Err(StdError::not_found("checkpoint"))
+        );
+        assert_eq!(
+            EVERY.may_load_at_height(&storage, DUMMY_KEY, 3),
+            Ok(Some(Some(100)))
+        );
+        assert_eq!(
```

Comment:
It would also be nice to test (only with EVERY) that `may_load_at_height(2)` will return the value set at 3 (the first change after that). That is important to assert the iterator works and not just exact matches.